File: backend/src/domain/identity/domain.rs
#![allow(dead_code)]
use std::sync::Arc;

use crate::{
    datasource::{dbdao::dao::DBDao, scylladao::dao::ScyllaDao, vectordao::dao::VectorDao},
    domain::identity::schema::{Permission, Role, User},
    infra::registry::Registry,
};

#[derive(Clone)]
pub struct IdentityDomain {
    pub db_dao: Arc<DBDao>,
    pub scylla_dao: Arc<ScyllaDao>,
    pub vector_dao: Arc<VectorDao>,
}

impl IdentityDomain {
    pub fn new(r: &Registry) -> Self {
        Self {
            db_dao: r.db_dao.clone(),
            scylla_dao: r.scylla_dao.clone(),
            vector_dao: r.vector_dao.clone(),
        }
    }

    pub async fn get_user(&self, id: i64) -> Option<User> {
        self.db_dao.get_user(id).await.ok().map(|record| User {
            id: record.id,
            username: record.username,
            email: record.email,
            password: record.password,
            role: record.role,
            enabled: Some(record.enabled),
            created_at: record.created_at,
        })
    }

    pub async fn search_users(
        &self,
        username: Option<String>,
        email: Option<String>,
        role: Option<String>,
        page: i64,
        limit: i64,
    ) -> (Vec<User>, i64) {
        match self
            .db_dao
            .search_users(username, email, role, page, limit)
            .await
        {
            Ok(records) => {
                let users: Vec<User> = records
                    .into_iter()
                    .map(|record| User {
                        id: record.id,
                        username: record.username,
                        email: record.email,
                        password: record.password,
                        role: record.role,
                        enabled: Some(record.enabled),
                        created_at: record.created_at,
                    })
                    .collect();
                let total = users.len() as i64;
                (users, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }

    pub async fn add_user(
        &self,
        username: String,
        email: String,
        password: Option<String>,
        role: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .add_user(username.as_str(), email.as_str(), password, role)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn update_user(
        &self,
        id: i64,
        username: Option<String>,
        email: Option<String>,
        password: Option<String>,
        role: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .update_user(id, username, email, password, role)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn get_role(&self, id: i64) -> Option<Role> {
        self.db_dao.get_role(id).await.ok().map(|record| Role {
            id: record.id,
            name: record.name,
            description: record.description,
            created_at: record.created_at,
        })
    }

    pub async fn search_roles(
        &self,
        name: Option<String>,
        description: Option<String>,
        page: i64,
        limit: i64,
    ) -> (Vec<Role>, i64) {
        match self
            .db_dao
            .search_roles(name, description, page, limit)
            .await
        {
            Ok(records) => {
                let roles: Vec<Role> = records
                    .into_iter()
                    .map(|record| Role {
                        id: record.id,
                        name: record.name,
                        description: record.description,
                        created_at: record.created_at,
                    })
                    .collect();
                let total = roles.len() as i64;
                (roles, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }

    pub async fn add_role(&self, name: String, description: Option<String>) -> Result<i64, String> {
        self.db_dao
            .add_role(name.as_str(), description)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn update_role(
        &self,
        id: i64,
        name: Option<String>,
        description: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .update_role(id, name, description)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn get_permission(&self, id: i64) -> Option<Permission> {
        self.db_dao
            .get_permission(id)
            .await
            .ok()
            .map(|record| Permission {
                id: record.id,
                name: record.name,
                description: record.description,
                created_at: record.created_at,
            })
    }

    pub async fn search_permissions(
        &self,
        name: Option<String>,
        description: Option<String>,
        page: i64,
        limit: i64,
    ) -> (Vec<Permission>, i64) {
        match self
            .db_dao
            .search_permissions(name, description, page, limit)
            .await
        {
            Ok(records) => {
                let permissions: Vec<Permission> = records
                    .into_iter()
                    .map(|record| Permission {
                        id: record.id,
                        name: record.name,
                        description: record.description,
                        created_at: record.created_at,
                    })
                    .collect();
                let total = permissions.len() as i64;
                (permissions, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }

    pub async fn add_permission(
        &self,
        name: String,
        description: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .add_permission(name.as_str(), description)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn update_permission(
        &self,
        id: i64,
        name: Option<String>,
        description: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .update_permission(id, name, description)
            .await
            .map_err(|e| e.to_string())
    }
}

File: backend/domain/identity/domain.go
package identity

import (
	"backend/datasource/dbdao"
	"backend/utils"
	"context"
	"errors"
)

type IdentityDomain struct {
	DB *dbdao.DB
}

func NewIdentityDomain(db *dbdao.DB) *IdentityDomain {
	return &IdentityDomain{DB: db}
}

func (d *IdentityDomain) GetUser(ctx context.Context, id int64) (*User, error) {
	user, err := d.DB.GetIdentityUser(ctx, id)
	if err != nil {
		return nil, err
	}
	return &User{
		ID:        user.ID,
		Username:  user.Username,
		Email:     user.Email,
		Password:  user.Password,
		Role:      user.Role,
		Enabled:   user.Enabled,
		CreatedAt: &user.CreatedAt,
	}, nil
}

func (d *IdentityDomain) SearchUsers(ctx context.Context, username, email, role string, page, limit int64) ([]*User, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}
	users, total, err := d.DB.SearchIdentityUsers(ctx, username, email, role, page, limit)
	if err != nil {
		return nil, 0, err
	}
	result := make([]*User, len(users))
	for i, u := range users {
		result[i] = &User{
			ID:        u.ID,
			Username:  u.Username,
			Email:     u.Email,
			Password:  u.Password,
			Role:      u.Role,
			Enabled:   u.Enabled,
			CreatedAt: &u.CreatedAt,
		}
	}
	return result, total, nil
}

func (d *IdentityDomain) AddUser(ctx context.Context, username, email, password, role string) (*User, error) {
	if username == "" || email == "" {
		return nil, errors.New("missing required fields")
	}
	user := &dbdao.IdentityUser{
		Username: username,
		Email:    email,
		Password: password,
		Role:     role,
		Enabled:  true,
	}
	created, err := d.DB.CreateIdentityUser(ctx, user)
	if err != nil {
		return nil, err
	}
	return &User{
		ID:        created.ID,
		Username:  created.Username,
		Email:     created.Email,
		Password:  created.Password,
		Role:      created.Role,
		Enabled:   created.Enabled,
		CreatedAt: &created.CreatedAt,
	}, nil
}

func (d *IdentityDomain) UpdateUser(ctx context.Context, id int64, username, email, password, role *string) (*User, error) {
	if id <= 0 {
		return nil, errors.New("invalid id")
	}
	user := &dbdao.IdentityUser{ID: id}
	if username != nil {
		user.Username = *username
	}
	if email != nil {
		user.Email = *email
	}
	if password != nil {
		user.Password = *password
	}
	if role != nil {
		user.Role = *role
	}
	updated, err := d.DB.UpdateIdentityUser(ctx, user)
	if err != nil {
		return nil, err
	}
	return &User{
		ID:        updated.ID,
		Username:  updated.Username,
		Email:     updated.Email,
		Password:  updated.Password,
		Role:      updated.Role,
		Enabled:   updated.Enabled,
		CreatedAt: &updated.CreatedAt,
	}, nil
}

func (d *IdentityDomain) GetRole(ctx context.Context, id int64) (*Role, error) {
	role, err := d.DB.GetRole(ctx, id)
	if err != nil {
		return nil, err
	}
	return &Role{
		ID:          role.ID,
		Name:        role.Name,
		Description: role.Description,
		CreatedAt:   &role.CreatedAt,
	}, nil
}

func (d *IdentityDomain) SearchRoles(ctx context.Context, name, description string, page, limit int64) ([]*Role, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}
	roles, total, err := d.DB.SearchRoles(ctx, name, description, page, limit)
	if err != nil {
		return nil, 0, err
	}
	result := make([]*Role, len(roles))
	for i, r := range roles {
		result[i] = &Role{
			ID:          r.ID,
			Name:        r.Name,
			Description: r.Description,
			CreatedAt:   &r.CreatedAt,
		}
	}
	return result, total, nil
}

func (d *IdentityDomain) AddRole(ctx context.Context, name, description string) (*Role, error) {
	if name == "" {
		return nil, errors.New("missing required fields")
	}
	role := &dbdao.Role{
		Name:        name,
		Description: description,
	}
	created, err := d.DB.CreateRole(ctx, role)
	if err != nil {
		return nil, err
	}
	return &Role{
		ID:          created.ID,
		Name:        created.Name,
		Description: created.Description,
		CreatedAt:   &created.CreatedAt,
	}, nil
}

func (d *IdentityDomain) UpdateRole(ctx context.Context, id int64, name, description *string) (*Role, error) {
	if id <= 0 {
		return nil, errors.New("invalid id")
	}
	role := &dbdao.Role{ID: id}
	if name != nil {
		role.Name = *name
	}
	if description != nil {
		role.Description = *description
	}
	updated, err := d.DB.UpdateRole(ctx, role)
	if err != nil {
		return nil, err
	}
	return &Role{
		ID:          updated.ID,
		Name:        updated.Name,
		Description: updated.Description,
		CreatedAt:   &updated.CreatedAt,
	}, nil
}

func (d *IdentityDomain) GetPermission(ctx context.Context, id int64) (*Permission, error) {
	perm, err := d.DB.GetPermission(ctx, id)
	if err != nil {
		return nil, err
	}
	return &Permission{
		ID:          perm.ID,
		Name:        perm.Name,
		Description: perm.Description,
		CreatedAt:   &perm.CreatedAt,
	}, nil
}

func (d *IdentityDomain) SearchPermissions(ctx context.Context, name, description string, page, limit int64) ([]*Permission, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}
	perms, total, err := d.DB.SearchPermissions(ctx, name, description, page, limit)
	if err != nil {
		return nil, 0, err
	}
	result := make([]*Permission, len(perms))
	for i, p := range perms {
		result[i] = &Permission{
			ID:          p.ID,
			Name:        p.Name,
			Description: p.Description,
			CreatedAt:   &p.CreatedAt,
		}
	}
	return result, total, nil
}

func (d *IdentityDomain) AddPermission(ctx context.Context, name, description string) (*Permission, error) {
	if name == "" {
		return nil, errors.New("missing required fields")
	}
	perm := &dbdao.Permission{
		Name:        name,
		Description: description,
	}
	created, err := d.DB.CreatePermission(ctx, perm)
	if err != nil {
		return nil, err
	}
	return &Permission{
		ID:          created.ID,
		Name:        created.Name,
		Description: created.Description,
		CreatedAt:   &created.CreatedAt,
	}, nil
}

func (d *IdentityDomain) UpdatePermission(ctx context.Context, id int64, name, description *string) (*Permission, error) {
	if id <= 0 {
		return nil, errors.New("invalid id")
	}
	perm := &dbdao.Permission{ID: id}
	if name != nil {
		perm.Name = *name
	}
	if description != nil {
		perm.Description = *description
	}
	updated, err := d.DB.UpdatePermission(ctx, perm)
	if err != nil {
		return nil, err
	}
	return &Permission{
		ID:          updated.ID,
		Name:        updated.Name,
		Description: updated.Description,
		CreatedAt:   &updated.CreatedAt,
	}, nil
}

// ===================== Authentication Methods =====================

// Login 用户登录
func (d *IdentityDomain) Login(ctx context.Context, username, password string) (*User, string, error) {
	// 查询用户
	users, _, err := d.DB.SearchIdentityUsers(ctx, username, "", "", 1, 1)
	if err != nil || len(users) == 0 {
		return nil, "", errors.New("user not found or password incorrect")
	}

	user := users[0]

	// 验证密码
	if !utils.VerifyPassword(user.Password, password) {
		return nil, "", errors.New("user not found or password incorrect")
	}

	// 生成 token
	token := utils.GenerateToken(user.ID, user.Email)

	return &User{
		ID:        user.ID,
		Username:  user.Username,
		Email:     user.Email,
		Password:  "", // 不返回密码哈希值
		Role:      user.Role,
		Enabled:   user.Enabled,
		CreatedAt: &user.CreatedAt,
	}, token, nil
}

// Register 用户注册
func (d *IdentityDomain) Register(ctx context.Context, username, email, password string) (*User, string, error) {
	// 检查用户名或邮箱是否已存在
	users, _, err := d.DB.SearchIdentityUsers(ctx, username, "", "", 1, 1)
	if err == nil && len(users) > 0 {
		return nil, "", errors.New("username already exists")
	}

	// 创建新用户
	hashedPassword := utils.HashPassword(password)
	id, err := d.DB.AddIdentityUser(ctx, &dbdao.IdentityUser{
		Username: username,
		Email:    email,
		Password: hashedPassword,
		Role:     "user",
		Enabled:  true,
	})
	if err != nil {
		return nil, "", err
	}

	// 获取刚创建的用户
	newUser, err := d.DB.GetIdentityUser(ctx, id)
	if err != nil {
		return nil, "", err
	}

	// 生成 token
	token := utils.GenerateToken(newUser.ID, newUser.Email)

	return &User{
		ID:        newUser.ID,
		Username:  newUser.Username,
		Email:     newUser.Email,
		Password:  "", // 不返回密码哈希值
		Role:      newUser.Role,
		Enabled:   newUser.Enabled,
		CreatedAt: &newUser.CreatedAt,
	}, token, nil
}
